fendermint/app/src/app.rs

```diff
@@ -151,6 +155,35 @@ where
             state_store: Arc::new(state_store),
             multi_engine: Arc::new(MultiEngine::new(1)),
             builtin_actors_bundle,
+            snapshot: None,
+            namespace: app_namespace,
+            state_hist: KVCollection::new(state_hist_namespace),
+            state_hist_size,
+            interpreter: Arc::new(interpreter),
+            exec_state: Arc::new(Mutex::new(None)),
+            check_state: Arc::new(tokio::sync::Mutex::new(None)),
+        };
+        app.init_committed_state()?;
+        Ok(app)
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    pub fn new_with_snapshot(
+        db: DB,
+        state_store: SS,
+        snapshot_config: SnapshotConfig,
+        builtin_actors_bundle: PathBuf,
+        app_namespace: S::Namespace,
+        state_hist_namespace: S::Namespace,
+        state_hist_size: u64,
+        interpreter: I,
+    ) -> Result<Self> {
+        let app = Self {
+            db: Arc::new(db),
+            state_store: Arc::new(state_store),
+            multi_engine: Arc::new(MultiEngine::new(1)),
+            builtin_actors_bundle,
+            snapshot: Some(SnapshotManager::new(snapshot_config)),
             namespace: app_namespace,
             state_hist: KVCollection::new(state_hist_namespace),
             state_hist_size,
```

Comment:
How about instead of repeating the whole constructor and potentially doing all sorts of combinations for future optional features, you leave `new` as it is, and add separate `with_snapshot` method that takes the snapshot config and initiates the manager, overwriting the previous one, if that is acceptable. It can consume an return `self`, so you can only do this during setup.

Reply:
^ +1